### DISCUSSION TITLE
Remove friction in minting - Automate transfers, sell offers/acceptance and reduce reserves by not holding 1000's of NFTs and offers.

### DISCUSSION THREAD

**Discussion by martypaz** [XRPLF/XRPL-Standards#99] [XLS Ideas (pre standard proposal)]
### Pain points

There a several pain points for users accepting mints from a marketplace. The goal here is to remove friction, improve speed of NFT delivery to the user’s wallet and reduce reserves.

As a user using our marketplace, you are able to win random NFTs from a collection using the spinner tool. We have many users who buy a lot of spins which results in a lot of time and effort – Prior to our bulk acceptance tool, accepting let’s say 20 NFT’s could take about 20 minutes.

From the xMart perspective, once an NFT has been allocated to a wallet, it’s then minted and the `NFTokenID` recorded. A 0 XRP sell offer is created for the user to accept. The user then goes to the offers and approvals page, ticks the items they want to accept. They then need to manually approve these in Xumm INDIVIDUALLY. This is not the fault of Xumm as multiple approvals could impact consumer protection.

### Improve NFTokenMinter options via automated transfers

My proposal is to improve the options available for the `NFTokenMinter`. This could allow us to Mint the token and specify a destination address as part of the payload or a flag to either Mint, Mint and create offer or Mint and transfer to wallet. 
This would remove the heavy burden of reserves for the minter as any offers created would automatically be accepted by the ledger on the buyer’s behalf. Plus we would have no reserves tied up in unclaimed tokens. 

The buyer would also benefit from a much faster delivery of the NFT without the need to approve it manually. The marketplaces would benefit from less XRP tied up in reserves and a more robust, single transaction process.

### Security

IMO, there would be no security risk as making it part of the mint process results in protection against subsequent offer creation. i.e., The minter would not be the owner. Additionally, as the buyer is already willing to accept NFTs from that collection/issuer and have paid for spins to do so it makes sense we can give them the NFT without the fuss.

I believe that as this is a transfer, secondary sales would not kick in as the price would be set at 0 XRP. Therefore, no royalties would be deducted.

This proposal should be easy to implement as it requires no new fields and simply expands on the functionality of the `NFTokenMinter`.

### Current implementation

`NFTokenMinter` can:

- Mint

### Proposed:

`NFTokenMinter` can:

- Mint
- Create initial offer during mint
- Transfer to Account without user needing to accept in their wallet

**Comment by WietseWind** [comment]
Based on user input, marketplace input, my own experience: yes, I think this should happen.

Looping in @nbougalis for a comment on feasibility of especially:
> Create initial offer during mint

**Comment by crtag** [comment]
There's indeed unnecessary friction in the NFT life cycle. This suggestion addresses the customer pain points and I support the idea. Though, I feel this approach alone might create more headaches in the long run such as spamming with NFTs and hence exhausting reserves etc. It feels that it's necessary to introduce a **preapproval** mechanism, as well as its **withdrawal** to allow certain addresses (`NFTokenMinter`) set to transfer into accounts w/o an offer (the last proposed point) that is the greatest benefit of the proposed update. 
Having said this, the existing "Authorized Minter" feature is what needs to get upgraded to allow authorized minters collection management instead.

**Reply by martypaz** [reply]
True, spamming is a concern (I myself have quite a few ETH NFTs that I never wanted) and I guess anyone could set a `NFTokenMinter` on their account root to purport being a valid NFT issuer. Although in our case, when a user sign in, perhaps there is a check we could do similar to checking minter status for an issuer. If we're not in the auto-accept whitelist then prompt them to approve us if they want immediate receipt of NFTS.

Now, the issue I had with `NFTokenMinter` is it's a single value. I imagine that was due to a lack of storage space on the account and having it as an array wasn't feasible. That meant a single wallet could only have xMart or only have marketplace xx as their minter.

I'm wondering if there is a way to attach a fungible token to an account whereby the auto-accept whitelist could have many entries. Or some other mechanism. I guess @nbougalis would know the answer.

**Reply by crtag** [reply]
Or, alternatively, manage the whitelist via NFTs. 
The critical point is to ensure the issuer remains a formal XRPL minter. Some places reduce friction by simply sending the mint transaction for signing by a "buyer" which is fine as long as it's simple IPFS, but totally wrong conceptually for a service. The proposal addresses this issue!

**Reply by martypaz** [reply]
I like the idea of the NFT whitelist. Then if the user wants to revoke access they/we can simply burn it. I assume I'm interpreting what you're saying correctly @crtag ?

**Reply by crtag** [reply]
Yep. That implies a bit of work how whitelists are accepted and recognized, etc. Basically, the whole life cycle. This does not necessary mean whitelisting by one way only, but it has to live on the ledger. Are there any similar concepts around? The idea would be to use NFT URI to hold either a reference to externally stored whitelist, ie on IPFS, which is not so cool as it relies on a third party, or, alternatively come up with a URI scheme and each NFT would hold a single whitelisted entity. Something like xrpl-minter-whitelist://rhgfujvdtyugfeyijvfujutghj

**Reply by crtag** [reply]
I gave it a bit of thinking and the better, and easier implementation of a "whitelist" can be achieved by introducing a new [NFT flag ](https://xrpl.org/nftoken.html#nftoken-flags), for example

**lsfAllowedMinter** 0xc000A
If used on NFT the URI content must contain a valid account address value. When an NFT minter attempts to "Transfer to Account without the recipient needing to accept the offer in their wallet" the minter address is validated against NFTs with the enabled **lsfAllowedMinter** flag.

This is a computational overhead and I'd assume should cause a higher transaction fee to compensate.

Also, a new transaction type is to be implemented **NFTokenMintTransfer** with this sort of payload

```
{
  "TransactionType": "NFTokenMintTransfer",
  "Account": "rvYAfWj5gh67oV6fW32ZzP3Aw4Eubs59B",
  "Destination" : "ra5nK24KXen9AHvsdFTKHSANinZseWnPcX",
  "TransferFee": 314,
  "NFTokenTaxon": 0,
  "Flags": 8,
  "Fee": "10",
  "URI": "697066733A2F2F62616679626569676479727A74357366703775646D37687537367568377932366E6634646675796C71616266336F636C67747179353566627A6469",
  "Memos": [
        {
            "Memo": {
                "MemoType":
                  "687474703A2F2F6578616D706C652E636F6D2F6D656D6F2F67656E65726963",
                "MemoData": "72656E74"
            }
        }
    ]
}
```
And in the case of a whitelisted minter would perform similarly to the normal minting process generating NFTTokenID with issuer address to ensure the NFToken URI content can be decoded as necessary.

**Reply by martypaz** [reply]
I like - Of course we still welcome some feedback from Nik

**Reply by mvadari** [reply]
How would you specify which accounts could mint on behalf of that account? A flag is just a boolean indicator.

**Reply by martypaz** [reply]
The idea here is that the flag would be on an NFT in the buyers wallet. When executing the NFTokenMintTransfer a lookup would be done to see if the use has a whitelist NFT that matches the minters wallet address. If so, allow to mint and transfer

**Reply by mvadari** [reply]
How would the whitelist be specified? It has to be stored on-ledger - it can't be stored on IPFS, because rippled doesn't have access to IPFS.
This is the most important aspect of this design, and is currently the least fleshed out.

**Reply by crtag** [reply]
The URI will contain the address to whitelist.

**Reply by mvadari** [reply]
It sounds like you're suggesting putting the whitelist in an NFT and storing it in IPFS. That won't work. rippled doesn't have any way to directly query IPFS - it doesn't have any network connection, and it almost certainly never will. The whitelist must be stored directly on-ledger somehow.

**Reply by crtag** [reply]
@mvadari No. The URI will contain the XRPL address to whitelist, how could you read it differently?

**Reply by crtag** [reply]
If there are 10 minter addresses to whitelist, the recipient would hold 10 NFTs with flag as described and each containing a single minter XRPL address.

**Reply by mvadari** [reply]
@crtag That will involve a lot of complexity for a transaction - rippled will need to iterate through all the NFTs the user owns (which could be in the millions), there's no way to easily filter the NFTs based on flags. You then need to iterate through all of the whitelist NFTs to determine whether the account is whitelisted.
This could perhaps be made easier if in the transaction you need to provide the NFTokenID to the whitelist NFT.

**Reply by crtag** [reply]
@mvadari brilliant idea. The minter has to do all of that and point to the NFT that whitelists them, slightly different transaction payload then.

**Reply by mvadari** [reply]
@crtag yes, but far more efficient

**Reply by crtag** [reply]
Something like this then, and `MinterValidation` must contain NFTokenID owned by `Destination` and composed as described above. All naming convention is up for some thinking. 

To be honest, this feels like one case of something more generic to implement, a kind of "permission" system that can be more flexible.

```
{
  "TransactionType": "NFTokenMintTransfer",
  "Account": "rvYAfWj5gh67oV6fW32ZzP3Aw4Eubs59B",
  "Destination" : "ra5nK24KXen9AHvsdFTKHSANinZseWnPcX",
  "MinterValidation": "<NFTokenID>"
  "TransferFee": 314,
  "NFTokenTaxon": 0,
  "Flags": 8,
  "Fee": "10",
  "URI": "697066733A2F2F62616679626569676479727A74357366703775646D37687537367568377932366E6634646675796C71616266336F636C67747179353566627A6469",
  "Memos": [
        {
            "Memo": {
                "MemoType":
                  "687474703A2F2F6578616D706C652E636F6D2F6D656D6F2F67656E65726963",
                "MemoData": "72656E74"
            }
        }
    ]
}
```

**Reply by ZachRoberts25** [reply]
This is a fantastic idea and would greatly improve the UX of XRPL NFTs.  Is anyone actively working on a solution to this in a PR somewhere I can reference.  I am happy to help out.  I wonder if this could be applied NFT Transfers as well.  Essentially NFT Buy Offers for 0 XRP.  Could we include this NFT whitelisting logic to those offers, and have them automatically accepted if the buyer/destination owns an whitelist NFT that states the Issuer on the receiving NFT is valid and can be accepted without the need to be signed.

**Reply by xmart-git** [reply]
I'm not sure it's actually being worked on Zach

**Comment by ajkagy** [comment]
I love the idea. I think there needs to be some kind of mechanism for a destination wallet to opt into the automatic transfer of an NFT from an authorized minter, otherwise you end up in a situation where this can be exploited by a malicious actor minting thousands of NFTs and automatically transferring the NFTs to the victims wallet thereby spamming wallets and locking reserves.

There's also a few questions of what happens to the NFT if the destination doesn't have the wallet reserves required for the transfer, is there a tec code through from the transaction by the ledger because part of the functionality failed considering the automatic transfer?

**Reply by martypaz** [reply]
See above comment for thoughts on the whitelist.

Again, Nik's input will be invaluable here about logic checks to ensure the transfer can go through. Then we handle that according to tecNoReserve etc Technically, we could check first their balance but it's another call in the process and accurately predicting reserves is not my bag. 

In fact I vaguely remember a discussion some time back between @WietseWind and I think Scott Schurr about reserves not being predictable or easily predictable at least.

[a comment by an automated account is not shown]
